### PR TITLE
Fix AzureStack swagger specs lint errors

### PR DESCRIPTION
**Pull request by AndyTao0402**
This change fixes the following lint errors for AzureStack swagger specs:

- ERROR (TrackedResourcePatchOperation/R3026/ARMViolation): Tracked resource 'Registration' must - have patch operation that at least supports the update of tags. It's strongly recommended that the PATCH operation supports update of all mutable properties as well.
    - file:///c:/Work/github/azure-rest-api-specs/specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Registration.json:214:8 ($.definitions.Registration)

- ERROR (TrackedResourcePatchOperation/R3026/ARMViolation): Tracked resource 'CustomerSubscription' must have patch operation that at least supports the update of tags. It's strongly recommended that the PATCH operation supports update of all mutable properties as well.
    - file:///c:/Work/github/azure-rest-api-specs/specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/CustomerSubscription.json:181:8 ($.definitions.CustomerSubscription)

- ERROR (TrackedResourcePatchOperation/R3026/ARMViolation): Tracked resource 'Product' must have patch operation that at least supports the update of tags. It's strongly recommended that the PATCH operation supports update of all mutable properties as well.
    - file:///c:/Work/github/azure-rest-api-specs/specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Product.json:314:8 ($.definitions.Product)

- ERROR (TrackedResourcePatchOperation/R3026/ARMViolation): Tracked resource 'RegistrationParameter' must have patch operation that at least supports the update of tags. It's strongly recommended that the PATCH operation supports update of all mutable properties as well.
    - file:///c:/Work/github/azure-rest-api-specs/specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Registration.json:275:8 ($.definitions.RegistrationParameter)

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [x] The title of the PR is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [X] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [X] If applicable, the PR references the bug/issue that it fixes.
- [X] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [X] My spec meets the review criteria:
  - [X] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [X] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [X] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

File: specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Registration.json [hunk 1]
```diff
@@ -171,6 +171,49 @@
                         }
                     }
                 }
+            },
+			"patch": {
+                "x-ms-examples": {
+                    "Create or update an Azure Stack registration.": {
+                        "$ref": "examples/Registration/Put.json"
```
Comment (annatisch) [inline specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Registration.json]: @AndyTao0402 - looks like it's referencing the wrong example?

Comment (AndyTao0402) [reply specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Registration.json]: True. Fixed as well as the description message in above line.

**Comment by annatisch** [comment]
Thanks @AndyTao0402 - the changes look okay to me, but are you aware that:
- There will be some breaking changes in any published SDKs
- The current readme.md will not generate a Python SDK
- The readme in the repo to generate the Node SDK is in the wrong location, so will not be detected.

Adding @ravbhatnagar to confirm the ARM changes.

**Comment by annatisch** [comment]
@AndyTao0402 - actually it looks like there's some model validation errors in the example, could you take a look?
```
 error: 
operationId: Registrations_Update
scenario: Patch an Azure Stack registration.
source: response
responseCode: '200'
severity: 0
errorCode: INVALID_TYPE
errorDetails:
  code: INVALID_TYPE
  params:
    - object
    - string
  message: Expected type object but found type string
  path: tags
  description: Custom tags for the resource.

 error: 
operationId: Registrations_Update
scenario: Patch an Azure Stack registration.
source: response
responseCode: '201'
severity: 0
errorCode: RESPONSE_SCHEMA_NOT_IN_SPEC
errorDetails:
  code: RESPONSE_SCHEMA_NOT_IN_SPEC
  id: OAV113
  message: >-
    Response statusCode "201" for operation "Registrations_Update" has response
    body provided in the example, however the response does not have a "schema"
    defined in the swagger spec.
  innerErrors: !<tag:yaml.org,2002:js/undefined> ''
  level: error

 error: 
operationId: Registrations_Update
scenario: Patch an Azure Stack registration.
source: response
responseCode: '404'
severity: 0
errorCode: RESPONSE_STATUS_CODE_NOT_IN_EXAMPLE
errorDetails:
  code: RESPONSE_STATUS_CODE_NOT_IN_EXAMPLE
  id: OAV111
  message: >-
    Following response status codes "404" for operation "Registrations_Update"
    were present in the swagger spec, however they were not present in
    x-ms-examples. Please provide them.
  innerErrors: !<tag:yaml.org,2002:js/undefined> ''
  level: error
```

**Comment by annatisch** [comment]
I haven't run the linter locally - I just look at the Travis CI run for the errors.
However I see that Travis is running this script for the validation:
`https://github.com/Azure/azure-rest-api-specs/blob/master/scripts/modelValidation.js`

Which in turn is calling into the node oav package. So you can probably run that directly :)

**Comment by AndyTao0402** [comment]
@annatisch, I have pushed a fix for example model validation. Can you check CI run to see if there's any other error?

**Comment by annatisch** [comment]
Thanks @AndyTao0402 - looks like you've added a syntax error - it's missing some `"` marks on line 21.

**Comment by AndyTao0402** [comment]
@annatisch Fixed.

**Comment by annatisch** [comment]
Thanks @AndyTao0402 - all the validation is passing now. :)

@ravbhatnagar - do you have any comments on the ARM violation fixes?

File: specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Registration.json [hunk 2]
```diff
@@ -171,6 +171,46 @@
...
+                    }
+                ],
+                "responses": {
+                    "200": {
+                        "description": "OK",
+                        "schema": {
+                            "$ref": "#/definitions/Registration"
+                        }
+                    },
+					"404": {
```
Comment (annatisch) [inline specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Registration.json]: @AndyTao0402 
Please remove all the error cases and add a "default" response which captures all error cases. The schema of the default response must comply with ARM error contract. sample can be found in Batch RP swagger.

**Comment by AndyTao0402** [comment]
@annatisch I have replaced all explicit error codes with a default response. Seems all test passed. Let me know if you have any further comments.

**Comment by annatisch** [comment]
Thanks @AndyTao0402!
Just a few example validation errors:
- This one still has a 404 response:
   https://github.com/AndyTao0402/azure-rest-api-specs/blob/801b6c3b4048d205edd10d0c3ba119b7ad94f332/specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/examples/Registration/List.json

- The "Product" definition uses "Resource" and not "TrackedResource" so does not have a "location" parameter:
   https://github.com/AndyTao0402/azure-rest-api-specs/blob/801b6c3b4048d205edd10d0c3ba119b7ad94f332/specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/examples/Product/Get.json
   https://github.com/AndyTao0402/azure-rest-api-specs/blob/801b6c3b4048d205edd10d0c3ba119b7ad94f332/specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/examples/Product/List.json

- The "CustomerSubscription" definition uses "Resource" and not "TrackedResource" so does not have a "location" parameter:
   https://github.com/AndyTao0402/azure-rest-api-specs/blob/801b6c3b4048d205edd10d0c3ba119b7ad94f332/specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/examples/CustomerSubscription/Put.json
   https://github.com/AndyTao0402/azure-rest-api-specs/blob/801b6c3b4048d205edd10d0c3ba119b7ad94f332/specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/examples/CustomerSubscription/Get.json
   https://github.com/AndyTao0402/azure-rest-api-specs/blob/801b6c3b4048d205edd10d0c3ba119b7ad94f332/specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/examples/CustomerSubscription/List.json

I'm not sure whether the service actually returns "location" in the response - which should determine whether the definitions need to be changed or the examples.

**Comment by AndyTao0402** [comment]
@annatisch We don't return location in Product and CustomerSubscription. Removed them from examples.

**Review by annatisch — APPROVED** [review]

**Comment by annatisch** [comment]
Thanks @AndyTao0402 
@ravbhatnagar hasn't confirmed the ARM changes yet, but given that it's been a few days I can merge if you would prefer not to wait.

**Comment by AndyTao0402** [comment]
@annatisch, we can wait as the new Patch operation change in our service hasn't been deployed yet.

File: specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Registration.json [hunk 3]
```diff
@@ -280,13 +332,24 @@
                     "description": "Properties of the Auzre Stack registration resource",
```
Comment (maxim-pop) [inline specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Registration.json]: "Azure" expected

Comment (maxim-pop) [inline specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Registration.json]: "Azure" expected

**Review by maxim-pop — CHANGES_REQUESTED** [review]
Misprint found

File: specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Registration.json (hunk 3 above)
Comment (annatisch) [reply specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Registration.json]: Nice catch :)

**Comment by jhendrixMSFT** [comment]
@AutorestCI regenerate azure-sdk-for-go

**Comment by jhendrixMSFT** [comment]
@AutorestCI regenerate azure-sdk-for-go

**Comment by jhendrixMSFT** [comment]
@AutorestCI regenerate azure-sdk-for-go

**Comment by ravbhatnagar** [comment]
This PR has more changes than just adding PATCH APIs. Some of those changes will result in breaking changes for the clients. Ex- adding new properties to request/response body. Is "location" a mandatory property? 
Are these just changes to swagger to get it in sync with service side? Or will these be actual changes on service side?

**Comment by AndyTao0402** [comment]
@ravbhatnagar, swagger specs are now in sync with service side. For Product and CustomerSubscription, they are proxy resources and location is not returned in response. For Registration, we added Patch operation in service side.

**Comment by AndyTao0402** [comment]
@annatisch, PATCH changes in service side has been deployed. You can merge now.

**Comment by annatisch** [comment]
Thanks @AndyTao0402 
@ravbhatnagar - are we good to go?

**Comment by AndyTao0402** [comment]
@ravbhatnagar, could you please review and sign off?

**Comment by ravbhatnagar** [comment]
So, after confirmation that this was just a swagger fix and service model was always like this, signing off on this PR>.

**Pull request merged**